File: easy/203_Remove_Linked_List_Elements.cpp
#include "../libstruct.h"

using namespace std;

/*
 * NO.203
 * Remove all elements from a linked list of integers that have value val.
 */

class Solution {
public:
    ListNode* removeElements(ListNode* head, int val) {
        ListNode **p = &head;
        while (*p) {
            if ((*p)->val == val)
                *p = (*p)->next;
            else
                p = &((*p)->next);
        }
        return head;
    }
};

/*
 * input: linked list: 1->2->3->2, and n = 2.
 * output: 1->3
 */
int main(int argc, char const *argv[])
{
    Solution solution;
    ListNode *head = new ListNode(1);
    head->next = new ListNode(2);
    head->next->next = new ListNode(3);
    head->next->next->next = new ListNode(2);
    solution.removeElements(head, 2);
    while(head) {cout << head->val << " "; head = head->next;}
    return 0;
}

File: easy/203_Remove_Linked_List_Elements.swift
//
//  203_Remove_Linked_List_Elements.swift
//  LeetCode
//
//  Created by wangzhe on 2018/7/27.
//  Copyright © 2018年 wangzhe. All rights reserved.
//

import Foundation

/*
 * NO.203
 * Remove all elements from a linked list of integers that have value val.
 */

public class ListNode {
    public var val: Int
    public var next: ListNode?
    public init(_ val: Int) {
        self.val = val
        self.next = nil
    }
}

class Solution {
    func removeElements(_ head: ListNode?, _ val: Int) -> ListNode? {
        let dummy = ListNode(0)
        dummy.next = head
        var node = dummy
        
        while node.next != nil {
            if node.next!.val == val {
                node.next = node.next!.next
            } else {
                node = node.next!
            }
        }
        
        return dummy.next
    }
}
